csrc/expr_evaluator.cpp

```diff
@@ -197,7 +197,35 @@ const PolymorphicValue& ExpressionEvaluator::evaluate(ParallelType pt) {
   return null_;
 }
 
-const PolymorphicValue& ExpressionEvaluator::getValue(const Val* value) {
+const PolymorphicValue& ExpressionEvaluator::evaluate(const Val* value) const {
+  if (precomputed_values_ && precomputed_values_->ready()) {
+    if (precomputed_values_->getMaybeValueFor(value).hasValue()) {
+      return precomputed_values_->getMaybeValueFor(value);
+    }
+  }
+
+  std::reference_wrapper<const PolymorphicValue> maybe_concrete_value =
+      getValue(value);
+  if (!maybe_concrete_value.get().hasValue()) {
+    if (auto def = value->definition()) {
+      FUSER_PERF_SCOPE("ExpressionEvaluator::evaluate");
+      std::vector<PolymorphicValue> inputs;
+      inputs.reserve(def->inputs().size());
+      for (auto i : def->inputs()) {
+        const auto& eval_i = evaluate(i);
+        if (!eval_i.hasValue()) {
+          return null_;
+        }
+        inputs.emplace_back(eval_i);
+      }
+      auto outputs = def->evaluate(*this, inputs);
+      maybe_concrete_value = getValue(value);
```

Comment:
I think this will ignore the recursion since `getValue` will look at `known_values_` which unlike in the non-const version has not been updated with `outputs` at this point. Instead I think we need to find the position of `value` in `outputs` and extract it here instead of using `getValue`. See my comment on the test for why this is currently passing.

Reply:
Should we do something like this?

```
const PolymorphicValue& ExpressionEvaluator::evaluateHelper(const Val* value, std::unordered_map<const Val*, PolymorphicValue>& known_values) const {
  recursively compute and write to known_values
}

const PolymorphicValue& ExpressionEvaluator::getValue(const Val* value, const std::unordered_map<const Val*, PolymorphicValue>& additional_known_values) const {
  find values in known_values or additional_known_values
}

const PolymorphicValue& ExpressionEvaluator::evaluate(const Val* value) {
  return evaluateHelper(value, known_values_);
}

PolymorphicValue ExpressionEvaluator::evaluate(const Val* value) {
  std::unordered_map<const Val*, PolymorphicValue> known_values;
  return evaluateHelper(value, known_values);
}
```

Reply:
Yeah that seems pretty clean. The only downside I can see is that in the const version we'd need to initialize `known_values` by copying `known_values_`.

Reply:
Oh now I see. You added `additional_known_values` to `getValue` so that `known_values` can be empty to start with. That's great. Then for the non-const `evaluate`, we can insert `additional_known_values` into `known_values_` after the call to `evaluateHelper`.

Reply:
> I think `evaluateHelper` should add to `additional_known_values`, otherwise, if you have one expr with multiple outputs, your recursion will call the same `evaluate` multiple times.

Yes exactly. I just meant that we should later save those new known values into `known_values_` in the non-recursive non-const `evaluate` call.